core/monitor/monitor.go

```diff
@@ -92,49 +93,104 @@ func New(mclient pb.MutationServiceClient,
-			if err := m.Process(mutResp); err != nil {
-				glog.Infof("Error processing mutations response: %v", err)
-			}
-		case err := <-errs:
-			// this is OK if there were no mutations in  between:
-			// TODO(ismail): handle the case when the known maxDuration has
-			// passed and no epoch was issued?
-			glog.Infof("Could not retrieve mutations API response %v", err)
+		case <-ctx.Done():
+			close(pairs)
+			return ctx.Err()
+		case pairs <- pair:
 		}
+		epochA = epoch
 	}
-
+	return nil
 }
 
-// Process processes a mutation response received from the keytransparency
-// server. Processing includes verifying, signing and storing the resulting
-// monitoring response.
-func (m *Monitor) Process(resp *pb.GetMutationsResponse) error {
-	var smr *trillian.SignedMapRoot
-	var err error
-	errs := m.VerifyMutationsResponse(resp)
-	if len(errs) == 0 {
-		glog.Infof("Successfully verified mutations response for epoch: %v", resp.Epoch)
-		smr, err = m.signMapRoot(resp)
+// ProcessLoop continuously fetches mutations and processes them.
+func (m *Monitor) ProcessLoop(ctx context.Context, domainID string, period time.Duration) error {
+	mutCli := mutationclient.New(m.mClient, period)
+	cctx, cancel := context.WithCancel(ctx)
+	errc := make(chan error)
+	epochs := make(chan *pb.Epoch)
+	pairs := make(chan EpochPair)
+
+	go func(ctx context.Context, domainID string, epochs chan<- *pb.Epoch) {
+		errc <- mutCli.StreamEpochs(ctx, domainID, 0, epochs)
+	}(cctx, domainID, epochs)
+	go func(ctx context.Context, epochs <-chan *pb.Epoch, pairs chan<- EpochPair) {
+		errc <- EpochPairs(ctx, epochs, pairs)
+	}(cctx, epochs, pairs)
+	defer cancel()
+
+	for pair := range pairs {
+		revision := pair.EpochB.GetSmr().GetMapRevision()
+		mutations, err := mutCli.EpochMutations(ctx, pair.EpochB)
 		if err != nil {
-			glog.Errorf("Failed to sign map root for epoch %v: %v", resp.Epoch, err)
-			return fmt.Errorf("m.signMapRoot(_): %v", err)
+			return err
 		}
+
+		var smr *trillian.SignedMapRoot
+		var errList []error
+		if errs := m.VerifyEpochMutations(pair.EpochA, pair.EpochB, mutations); len(errs) > 0 {
+			glog.Infof("Epoch %v did not verify: %v", revision, errs)
+			errList = errs
+		} else {
+			// Sign if successful.
+			smr, err = m.signMapRoot(pair.EpochB.GetSmr())
+			if err != nil {
+				return err
+			}
+		}
+
+		// Save result.
+		if err := m.store.Set(revision, &monitorstorage.Result{
+			Smr:    smr,
+			Seen:   time.Now(),
+			Errors: errList,
+		}); err != nil {
+			return fmt.Errorf("monitorstorage.Set(%v, _): %v", revision, err)
+		}
+		return nil
+
 	}
-	if err := m.store.Set(resp.Epoch, &monitorstorage.Result{
-		Smr:      smr,
-		Seen:     time.Now(),
-		Errors:   errs,
-		Response: resp,
-	}); err != nil {
-		return fmt.Errorf("monitorstorage.Set(%v, _): %v", resp.Epoch, err)
+	errA := <-errc
+	errB := <-errc
+	if errA != nil || errB != nil {
+		return fmt.Errorf("failed fetching epochs: (fetch: %v, pair: %v)", errA, errB)
```

Comment:
Is it racy to assume that `errA` and `errB` will be in this particular order?
Could the pairing go-routine emit an error to `errc` terminate before the fetching go-routine does?